heat/core/dndarray.py

```diff
@@ -1464,6 +1537,59 @@ def __gt__(self, other):
         """
         return relational.gt(self, other)
 
+    def halorize_(self):
+        """
+        Exchange halo data between ranks
+        """
+        if self.is_distributed() and not self.is_halorized():
+            send_before = self.__prephalo(0, self.halo)
+            send_after = self.__prephalo(-self.halo, None)
+
+            shape = tuple(send_before.shape)
+
+            recv_before = None
+            recv_after = None
+
+            if self.comm.rank != self.comm.size - 1:
+                self.comm.Isend(send_after, self.comm.rank + 1)
+                recv_after = torch.zeros(
+                    shape, dtype=send_before.dtype, device=self.device.torch_device
+                )
+                self.comm.Recv(recv_after, self.comm.rank + 1)
+
+            if self.comm.rank != 0:
+                self.comm.Isend(send_before, self.comm.rank - 1)
+                recv_before = torch.zeros(
+                    shape, dtype=send_after.dtype, device=self.device.torch_device
+                )
+                self.comm.Recv(recv_before, self.comm.rank - 1)
+
+            self.__halos = [recv_before, recv_after]
+
+        return
+
+    def halorized(self):
+        """
+        Returns
+        -------
+            tensor: concatenation of halo_before, array and halo_after along the split axis
+        """
+        if not self.is_halorized():
+            return self
+
+        cat = []
+        if self.__halos[0] is not None:
+            cat.append(self.__halos[0])
+
+        cat.append(self.__array)
+
+        if self.__halos[1] is not None:
+            cat.append(self.__halos[1])
+
+        out = torch.cat(cat, self.split)
+
+        return DNDarray(out, out.shape, self.dtype, None, self.device, None, None)
```

Comment:
This is a dangerous thing to do. The new array content would be now the actual array plus the catted halos on the side. Some low-level code will directly access `__array` and will not get all data including halos. I believe it would be good to introduce a new member that is called something along the lines of `__raw_data`, `__storage`,` __array_with_halos` and make `__array` to be a property that will return the actual array content.